File: samples/language-impl-comparison/fsharp/json_parser.fs
// JSON パーサー (F# 実装)
// JSON 構文を解析して汎用値型に変換する

module JsonParser

open System
open System.Collections.Generic

/// JSON 値型
type JsonValue =
  | JNull
  | JBool of bool
  | JNumber of float
  | JString of string
  | JArray of JsonValue list
  | JObject of Map<string, JsonValue>

/// トークン型
type Token =
  | LBrace
  | RBrace
  | LBracket
  | RBracket
  | Colon
  | Comma
  | StringLiteral of string
  | NumberLiteral of float
  | BoolLiteral of bool
  | NullLiteral

/// パース状態
type ParseState = { Tokens: Token list }

/// パースエラー
type ParseError =
  | UnexpectedEOF
  | UnexpectedToken of expected: string * found: Token

/// トークン化
let tokenize (source: string) : Token list =
  let rec loop (index: int) (acc: Token list) =
    if index >= source.Length then
      List.rev acc
    else
      let ch = source.[index]
      match ch with
      | ' ' | '\n' | '\t' | '\r' -> loop (index + 1) acc
      | '{' -> loop (index + 1) (LBrace :: acc)
      | '}' -> loop (index + 1) (RBrace :: acc)
      | '[' -> loop (index + 1) (LBracket :: acc)
      | ']' -> loop (index + 1) (RBracket :: acc)
      | ':' -> loop (index + 1) (Colon :: acc)
      | ',' -> loop (index + 1) (Comma :: acc)
      | 't' ->
          if source.Substring(index, 4) = "true" then
            loop (index + 4) (BoolLiteral true :: acc)
          else
            loop (index + 1) acc
      | 'f' ->
          if source.Substring(index, 5) = "false" then
            loop (index + 5) (BoolLiteral false :: acc)
          else
            loop (index + 1) acc
      | 'n' ->
          if source.Substring(index, 4) = "null" then
            loop (index + 4) (NullLiteral :: acc)
          else
            loop (index + 1) acc
      | '"' ->
          let endIndex = source.IndexOf('"', index + 1)
          let str = source.Substring(index + 1, endIndex - index - 1)
          loop (endIndex + 1) (StringLiteral str :: acc)
      | _ ->
          // 数値の読み取り (簡易実装)
          let mutable endIndex = index
          while endIndex < source.Length &&
                (Char.IsDigit(source.[endIndex]) ||
                 source.[endIndex] = '.' ||
                 source.[endIndex] = '-') do
            endIndex <- endIndex + 1
          let numStr = source.Substring(index, endIndex - index)
          match Double.TryParse(numStr) with
          | (true, num) -> loop endIndex (NumberLiteral num :: acc)
          | _ -> loop (index + 1) acc
  loop 0 []

/// 値のパース
let rec parseValue (state: ParseState) : Result<JsonValue * ParseState, ParseError> =
  match state.Tokens with
  | [] -> Error UnexpectedEOF
  | token :: rest ->
      match token with
      | NullLiteral -> Ok (JNull, { Tokens = rest })
      | BoolLiteral flag -> Ok (JBool flag, { Tokens = rest })
      | NumberLiteral num -> Ok (JNumber num, { Tokens = rest })
      | StringLiteral text -> Ok (JString text, { Tokens = rest })
      | LBracket -> parseArray { Tokens = rest }
      | LBrace -> parseObject { Tokens = rest }
      | other -> Error (UnexpectedToken ("値", other))

/// 配列のパース
and parseArray (state: ParseState) : Result<JsonValue * ParseState, ParseError> =
  match state.Tokens with
  | RBracket :: rest -> Ok (JArray [], { Tokens = rest })
  | _ ->
      let rec loop (current: ParseState) (acc: JsonValue list) =
        match parseValue current with
        | Error err -> Error err
        | Ok (value, next) ->
            let newAcc = acc @ [value]
            match next.Tokens with
            | Comma :: rest -> loop { Tokens = rest } newAcc
            | RBracket :: rest -> Ok (JArray newAcc, { Tokens = rest })
            | token :: _ -> Error (UnexpectedToken ("]", token))
            | [] -> Error UnexpectedEOF
      loop state []

/// オブジェクトのパース
and parseObject (state: ParseState) : Result<JsonValue * ParseState, ParseError> =
  match state.Tokens with
  | RBrace :: rest -> Ok (JObject Map.empty, { Tokens = rest })
  | _ ->
      let rec loop (current: ParseState) (acc: Map<string, JsonValue>) =
        match current.Tokens with
        | StringLiteral key :: Colon :: rest ->
            match parseValue { Tokens = rest } with
            | Error err -> Error err
            | Ok (value, next) ->
                let newAcc = Map.add key value acc
                match next.Tokens with
                | Comma :: rest' -> loop { Tokens = rest' } newAcc
                | RBrace :: rest' -> Ok (JObject newAcc, { Tokens = rest' })
                | token :: _ -> Error (UnexpectedToken ("}", token))
                | [] -> Error UnexpectedEOF
        | token :: _ -> Error (UnexpectedToken ("文字列", token))
        | [] -> Error UnexpectedEOF
      loop state Map.empty

/// メインパース関数
let parseJson (source: string) : Result<JsonValue, string> =
  let tokens = tokenize source
  let state = { Tokens = tokens }
  match parseValue state with
  | Error UnexpectedEOF -> Error "予期しない入力終端"
  | Error (UnexpectedToken (expected, found)) ->
      Error $"期待: {expected}, 実際: {found}"
  | Ok (value, rest) ->
      if List.isEmpty rest.Tokens then
        Ok value
      else
        Error "末尾に未消費トークンがあります"

// 利用例
// parseJson """{"name": "Alice", "age": 30}"""
// => Ok (JObject (map [("name", JString "Alice"); ("age", JNumber 30.0)]))

File: samples/language-impl-comparison/swift/json_parser.swift
// JSON パーサー (Swift 実装)
// JSON 構文を解析して汎用値型に変換する

import Foundation

// JSON 値型
enum JsonValue {
  case jNull
  case jBool(Bool)
  case jNumber(Double)
  case jString(String)
  case jArray([JsonValue])
  case jObject([String: JsonValue])
}

// トークン型
enum Token {
  case lBrace
  case rBrace
  case lBracket
  case rBracket
  case colon
  case comma
  case stringLiteral(String)
  case numberLiteral(Double)
  case boolLiteral(Bool)
  case nullLiteral
}

// パース状態
struct ParseState {
  let tokens: [Token]
}

// パースエラー
enum ParseError: Error {
  case unexpectedEOF
  case unexpectedToken(expected: String, found: Token)
}

// トークン化
func tokenize(_ source: String) -> [Token] {
  func loop(index: Int, acc: [Token]) -> [Token] {
    guard index < source.count else {
      return acc.reversed()
    }

    let ch = source[source.index(source.startIndex, offsetBy: index)]

    switch ch {
    case " ", "\n", "\t", "\r":
      return loop(index: index + 1, acc: acc)
    case "{":
      return loop(index: index + 1, acc: [.lBrace] + acc)
    case "}":
      return loop(index: index + 1, acc: [.rBrace] + acc)
    case "[":
      return loop(index: index + 1, acc: [.lBracket] + acc)
    case "]":
      return loop(index: index + 1, acc: [.rBracket] + acc)
    case ":":
      return loop(index: index + 1, acc: [.colon] + acc)
    case ",":
      return loop(index: index + 1, acc: [.comma] + acc)
    case "t":
      let startIdx = source.index(source.startIndex, offsetBy: index)
      let endIdx = source.index(startIdx, offsetBy: 4, limitedBy: source.endIndex) ?? source.endIndex
      if source[startIdx..<endIdx] == "true" {
        return loop(index: index + 4, acc: [.boolLiteral(true)] + acc)
      }
      return loop(index: index + 1, acc: acc)
    case "f":
      let startIdx = source.index(source.startIndex, offsetBy: index)
      let endIdx = source.index(startIdx, offsetBy: 5, limitedBy: source.endIndex) ?? source.endIndex
      if source[startIdx..<endIdx] == "false" {
        return loop(index: index + 5, acc: [.boolLiteral(false)] + acc)
      }
      return loop(index: index + 1, acc: acc)
    case "n":
      let startIdx = source.index(source.startIndex, offsetBy: index)
      let endIdx = source.index(startIdx, offsetBy: 4, limitedBy: source.endIndex) ?? source.endIndex
      if source[startIdx..<endIdx] == "null" {
        return loop(index: index + 4, acc: [.nullLiteral] + acc)
      }
      return loop(index: index + 1, acc: acc)
    case "\"":
      let startIdx = source.index(source.startIndex, offsetBy: index + 1)
      if let endIdx = source[startIdx...].firstIndex(of: "\"") {
        let str = String(source[startIdx..<endIdx])
        let nextIndex = source.distance(from: source.startIndex, to: endIdx) + 1
        return loop(index: nextIndex, acc: [.stringLiteral(str)] + acc)
      }
      return loop(index: index + 1, acc: acc)
    default:
      // 数値の読み取り (簡易実装)
      var endIndex = index
      while endIndex < source.count {
        let ch = source[source.index(source.startIndex, offsetBy: endIndex)]
        if ch.isNumber || ch == "." || ch == "-" {
          endIndex += 1
        } else {
          break
        }
      }
      let startIdx = source.index(source.startIndex, offsetBy: index)
      let endIdx = source.index(source.startIndex, offsetBy: endIndex)
      let numStr = String(source[startIdx..<endIdx])
      if let num = Double(numStr) {
        return loop(index: endIndex, acc: [.numberLiteral(num)] + acc)
      }
      return loop(index: index + 1, acc: acc)
    }
  }
  return loop(index: 0, acc: [])
}

// 値のパース
func parseValue(_ state: ParseState) -> Result<(JsonValue, ParseState), ParseError> {
  guard !state.tokens.isEmpty else {
    return .failure(.unexpectedEOF)
  }
  let token = state.tokens[0]
  let rest = Array(state.tokens.dropFirst())

  switch token {
  case .nullLiteral:
    return .success((.jNull, ParseState(tokens: rest)))
  case .boolLiteral(let flag):
    return .success((.jBool(flag), ParseState(tokens: rest)))
  case .numberLiteral(let num):
    return .success((.jNumber(num), ParseState(tokens: rest)))
  case .stringLiteral(let text):
    return .success((.jString(text), ParseState(tokens: rest)))
  case .lBracket:
    return parseArray(ParseState(tokens: rest))
  case .lBrace:
    return parseObject(ParseState(tokens: rest))
  default:
    return .failure(.unexpectedToken(expected: "値", found: token))
  }
}

// 配列のパース
func parseArray(_ state: ParseState) -> Result<(JsonValue, ParseState), ParseError> {
  if !state.tokens.isEmpty && state.tokens[0] == .rBracket {
    let rest = Array(state.tokens.dropFirst())
    return .success((.jArray([]), ParseState(tokens: rest)))
  }

  func loop(current: ParseState, acc: [JsonValue]) -> Result<(JsonValue, ParseState), ParseError> {
    switch parseValue(current) {
    case .failure(let err):
      return .failure(err)
    case .success(let (value, next)):
      let newAcc = acc + [value]
      guard !next.tokens.isEmpty else {
        return .failure(.unexpectedEOF)
      }
      let token = next.tokens[0]
      let rest = Array(next.tokens.dropFirst())
      switch token {
      case .comma:
        return loop(current: ParseState(tokens: rest), acc: newAcc)
      case .rBracket:
        return .success((.jArray(newAcc), ParseState(tokens: rest)))
      default:
        return .failure(.unexpectedToken(expected: "]", found: token))
      }
    }
  }
  return loop(current: state, acc: [])
}

// オブジェクトのパース
func parseObject(_ state: ParseState) -> Result<(JsonValue, ParseState), ParseError> {
  if !state.tokens.isEmpty && state.tokens[0] == .rBrace {
    let rest = Array(state.tokens.dropFirst())
    return .success((.jObject([:]), ParseState(tokens: rest)))
  }

  func loop(current: ParseState, acc: [String: JsonValue]) -> Result<(JsonValue, ParseState), ParseError> {
    guard current.tokens.count >= 2 else {
      return .failure(.unexpectedEOF)
    }
    guard case .stringLiteral(let key) = current.tokens[0],
          case .colon = current.tokens[1] else {
      return .failure(.unexpectedToken(expected: "文字列:値", found: current.tokens[0]))
    }
    let rest = Array(current.tokens.dropFirst(2))

    switch parseValue(ParseState(tokens: rest)) {
    case .failure(let err):
      return .failure(err)
    case .success(let (value, next)):
      var newAcc = acc
      newAcc[key] = value
      guard !next.tokens.isEmpty else {
        return .failure(.unexpectedEOF)
      }
      let token = next.tokens[0]
      let rest2 = Array(next.tokens.dropFirst())
      switch token {
      case .comma:
        return loop(current: ParseState(tokens: rest2), acc: newAcc)
      case .rBrace:
        return .success((.jObject(newAcc), ParseState(tokens: rest2)))
      default:
        return .failure(.unexpectedToken(expected: "}", found: token))
      }
    }
  }
  return loop(current: state, acc: [:])
}

// メインパース関数
func parseJson(_ source: String) -> Result<JsonValue, String> {
  let tokens = tokenize(source)
  let state = ParseState(tokens: tokens)
  switch parseValue(state) {
  case .failure(.unexpectedEOF):
    return .failure("予期しない入力終端")
  case .failure(.unexpectedToken(let expected, _)):
    return .failure("期待: \(expected)")
  case .success(let (value, rest)):
    if rest.tokens.isEmpty {
      return .success(value)
    } else {
      return .failure("末尾に未消費トークンがあります")
    }
  }
}

// 利用例
// parseJson("""{"name": "Alice", "age": 30}""")
// => .success(.jObject(["name": .jString("Alice"), "age": .jNumber(30.0)]))